### PR TITLE
PAYMENTS-4759 Add support to get and filter instruments

### PR DESCRIPTION
**Pull request by icatalina**
## What?
- Add support for account instruments
- Add support to filter instruments based on payment method type

## Why?
- We want to support account-type instruments required for paypal vaulting
- We would like to future proof the SDK so new instruments added to the API don't break old copies of the Checkout SDK.
- **From now onwards:** In order to instrument types that are not of type `card` you need to provide the `paymentMethod` to `getAllInstruments`

## Testing / Proof
- Unit / Functional

@bigcommerce/checkout @bigcommerce/payments

### CODE REVIEW

File: src/payment/instrument/instrument-response-body.ts [hunk 1]
```diff
@@ -5,16 +5,27 @@ export interface InstrumentError {
     message: string;
 }
 
-export interface InternalInstrument {
+export  type InternalInstrument = CardInternalInstrument | AccountInternalInstrument;
```
Comment (davidchin) [inline src/payment/instrument/instrument-response-body.ts]: Extra space in front of `type` 💅

File: src/payment/instrument/instrument-selector.ts [hunk 2]
```diff
@@ -19,7 +22,25 @@ export type InstrumentSelectorFactory = (state: InstrumentState) => InstrumentSe
 export function createInstrumentSelectorFactory(): InstrumentSelectorFactory {
     const getInstruments = createSelector(
         (state: InstrumentState) => state.data,
-        instruments => () => instruments
+        instruments => (paymentMethod?: PaymentMethod) => {
+            if (!instruments) {
+                return undefined;
```
Comment (davidchin) [inline src/payment/instrument/instrument-selector.ts]: Just `return` is enough 💅

File: src/payment/instrument/supported-payment-instruments.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import Instrument from './instrument';
+
+interface SupportedInstruments {
+    [key: string]: Partial<Instrument>;
```
Comment (davidchin) [inline src/payment/instrument/supported-payment-instruments.ts]: Maybe `Pick<Instrument, 'provider' | 'method'>` is more accurate so `provider` and `method` are not marked as optional.

File: src/payment/instrument/instrument-selector.ts [hunk 4]
```diff
@@ -19,7 +22,25 @@ export type InstrumentSelectorFactory = (state: InstrumentState) => InstrumentSe
...
+            }
+
+            if (!paymentMethod) {
+                const supportedCardInstruments = filter(supportedInstruments, { method: 'card' });
+
+                return flatMap(supportedCardInstruments, cardInstrument => filter(instruments, cardInstrument));
```
Comment (davidchin) [inline src/payment/instrument/instrument-selector.ts]: Is the order of `instruments` important? i.e.: Right now, it seems the order of `Instrument` is based on `supportedInstruments` and then `instruments`. If it's not important, then what you have here is good :)

Comment (icatalina) [reply src/payment/instrument/instrument-selector.ts]: We only show one instrument type at a given time, so as long as we retain the order for every type of instrument it should be ok.

**Pull request merged**